File: apps/api/src/carpool_api/schemas/optimizations.py
"""Optimization job payloads (docs/design.md 6, 4.2)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict

from carpool_api.models import JobStatus, OptimizationJob
from carpool_api.schemas.weights import Weights

#: The only algorithm a user request may ask for.
#:
#: CP-SAT is deliberately absent: it is a measuring instrument, not a product feature. It costs
#: ~400 MB per solve on a 2 GB host, its runtime varies wildly, and the 50-participant cap sits past
#: the ~40 where it reliably proves optimality -- so it runs in the benchmark harness and never on a
#: user request, which is also what keeps OR-Tools out of the production image (docs/design.md 8.3).
#: LNS joins this list in Week 4.
Algorithm = Literal["greedy"]


class OptimizationCreate(BaseModel):
    """What to solve and how.

    No `preserve_previous` yet, though docs/design.md 6 lists one: it would set the churn weight
    against the previously activated solution, and greedy cannot steer by churn: it constructs
    from nothing. The flag arrives with LNS, whose acceptance criterion can actually honour it
    (docs/design.md 8.4). Offering it now would be a switch that silently does nothing.
    """

    model_config = ConfigDict(extra="forbid")

    algorithm: Algorithm = "greedy"
    #: Overrides `events.settings` for this job only. The effective set is recorded in
    #: `optimization_jobs.params`, so the job stays reproducible after the event's settings change.
    weights: Weights | None = None


class JobRead(BaseModel):
    """A job as the client polls it (docs/design.md 4.3 -- polling, not SSE, in v1)."""

    job_id: uuid.UUID
    event_public_id: str
    status: JobStatus
    algorithm: str
    #: The weights this job actually ran with, resolved from request and event settings.
    weights: Weights
    #: `events.participants_version` as read at enqueue time. A solution whose version is behind the
    #: event's current one describes a roster that has since changed (docs/design.md 5.1).
    input_version: int
    progress: dict[str, Any] | None
    error: str | None
    #: Present once the job has succeeded.
    solution_id: uuid.UUID | None
    cancel_requested: bool
    queued_at: datetime
    started_at: datetime | None
    finished_at: datetime | None

    @classmethod
    def of(
        cls, job: OptimizationJob, *, event_public_id: str, solution_id: uuid.UUID | None
    ) -> Self:
        return cls(
            job_id=job.id,
            event_public_id=event_public_id,
            status=job.status,
            algorithm=job.algorithm,
            weights=Weights.model_validate(job.params.get("weights") or {}),
            input_version=job.input_version,
            progress=job.progress,
            error=job.error,
            solution_id=solution_id,
            cancel_requested=job.cancel_requested,
            queued_at=job.queued_at,
            started_at=job.started_at,
            finished_at=job.finished_at,
        )


class JobConflict(BaseModel):
    """The 409 body when a job is already in flight for this event.

    Carries the existing job id so the client polls that one instead of retrying -- a retry would
    hit the same partial unique index and fail again (docs/design.md 5.1).
    """

    detail: str
    #: Null in the narrow case where the job that won the race finished between the failed insert
    #: and the lookup. The conflict was still real; there is simply nothing left to poll, and the
    #: client should just try again.
    existing_job_id: uuid.UUID | None

File: apps/web/lib/api/optimizations.ts
/**
 * Optimization jobs.
 *
 * Written against the *async* contract -- enqueue, then poll until terminal -- even though the API
 * currently solves inline and hands back a finished job in the POST response. That is a deliberate
 * seam: `docs/design.md` 4.2 makes a separate worker a Week 4 decision, and a client that assumes
 * the answer arrives with the 202 is a client that breaks the day one lands. Polling a job that is
 * already `succeeded` costs exactly one request.
 */

import { api, describeError, organizerAuth, unwrap } from "./client";
import type { components } from "./schema";

export type Job = components["schemas"]["JobRead"];
export type JobStatus = Job["status"];
export type Weights = components["schemas"]["Weights"];

/** Statuses a job can still move out of. Anything else is final and stops the polling. */
const IN_FLIGHT: readonly JobStatus[] = ["queued", "running"];

export function isTerminal(status: JobStatus): boolean {
  return !IN_FLIGHT.includes(status);
}

/**
 * The outcome of asking for an optimization.
 *
 * `conflict` is not an error to show and forget: the API answers 409 when another job is already in
 * flight for this event, and hands back the id of the job that won so the client can poll *that*
 * one instead. Retrying would hit the same partial unique index and fail again.
 */
export type StartResult =
  | { kind: "job"; job: Job }
  | { kind: "conflict"; existingJobId: string | null; detail: string };

type JobConflict = components["schemas"]["JobConflict"];

/**
 * Ask for a solve.
 *
 * 202 means a job was created; 200 means an existing one is being returned -- an idempotent replay,
 * or a fingerprint that was already solved. The body is identical, so this does not distinguish
 * them; the caller polls `job_id` either way.
 */
export async function startOptimization(publicId: string): Promise<StartResult> {
  const result = await api.POST("/v1/events/{public_id}/optimizations", {
    params: { path: { public_id: publicId }, header: undefined },
    headers: organizerAuth(publicId),
    body: { algorithm: "greedy" },
  });

  if (result.response.status === 409) {
    const body = result.error as JobConflict | undefined;
    return {
      kind: "conflict",
      existingJobId: body?.existing_job_id ?? null,
      detail: body?.detail ?? "An optimization is already running for this event.",
    };
  }
  if (result.error !== undefined || result.data === undefined) {
    throw describeError(result.response.status, result.error);
  }
  return { kind: "job", job: result.data };
}

export async function fetchJob(publicId: string, jobId: string): Promise<Job> {
  return unwrap(
    await api.GET("/v1/events/{public_id}/optimizations/{job_id}", {
      params: { path: { public_id: publicId, job_id: jobId }, header: undefined },
      headers: organizerAuth(publicId),
    }),
  );
}

/** Cooperative cancel: the solver checks the flag, so a running job stops at its next look. */
export async function cancelJob(publicId: string, jobId: string): Promise<Job> {
  return unwrap(
    await api.DELETE("/v1/events/{public_id}/optimizations/{job_id}", {
      params: { path: { public_id: publicId, job_id: jobId }, header: undefined },
      headers: organizerAuth(publicId),
    }),
  );
}

export const jobKeys = {
  detail: (publicId: string, jobId: string) => ["job", publicId, jobId] as const,
};

/** What to tell the organizer while a job is in flight, or about how it ended. */
export function describeJob(job: Job): string {
  switch (job.status) {
    case "queued":
      return "Queued…";
    case "running":
      return "Solving…";
    case "succeeded":
      return "Done.";
    case "cancelled":
      return "Cancelled.";
    case "failed":
      return job.error ? `Failed: ${job.error}` : "The solve failed.";
    default:
      return job.status;
  }
}
